nilearn/decomposition/base.py

```diff
@@ -0,0 +1,469 @@
+        This parameter is passed to signal.clean. Please see the
+        corresponding documentation for details.
+
+    reduction_ratio: 'auto' or float in [0., 1.], optional
+        - Between 0. or 1. : controls compression of data, 1. means no
+        compression
+        - if set to 'auto', estimator will set the number of components per
+        compressed session to be n_components
+
+    n_components: integer, optional
+        Number of components to be extracted by the PCA
+
+    random_state: int or RandomState
+        Pseudo number generator state used for random sampling.
+
+    memory_level: integer, optional
+        Integer indicating the level of memorization. The higher, the more
+        function calls are cached.
+
+    memory: joblib.Memory
+        Used to cache the function calls.
+
+    Retuns
+    ------
+    data: ndarray or memorymap
+        Concatenation of reduced data
+    """
+
+    if not hasattr(imgs, '__iter__'):
+        imgs = [imgs]
+    else:
+        imgs = imgs
+
+    if reduction_ratio == 'auto':
+        if n_components is None:
+            # Reduction ratio is 1 if
+            # neither n_components nor ratio is provided
+            reduction_ratio = 1
+        else:
+            reduction_ratio = 'auto'
+    else:
+        if reduction_ratio is None:
+            reduction_ratio = 1
+        else:
+            reduction_ratio = float(reduction_ratio)
+        if not 0 <= reduction_ratio <= 1:
+            raise ValueError('Reduction ratio should be between 0., 1.,'
+                             'got %.2f' % reduction_ratio)
+
+    if confounds is None:
+        confounds = [None] * len(imgs)
+
+    # Precomputing number of samples for preallocation
+    subject_n_samples = np.zeros(len(imgs), dtype='int')
+    for i, img in enumerate(imgs):
+        this_n_samples = check_niimg_4d(img).shape[3]
+        if reduction_ratio == 'auto':
+            subject_n_samples[i] = min(n_components,
+                                       this_n_samples)
+        else:
+            subject_n_samples[i] = int(ceil(this_n_samples *
+                                            reduction_ratio))
+    n_voxels = np.sum(_safe_get_data(masker.mask_img_))
+    n_samples = np.sum(subject_n_samples)
+
+    # XXX Should we provided memory mapping for n_jobs > 1 to allow concurrent
+    # write ?
+    data = np.empty((n_samples, n_voxels), order='F',
+                    dtype='float64')
+
+    data_list = Parallel(n_jobs=n_jobs)(
+        delayed(_mask_and_reduce_single)(
+            masker,
+            img, confound,
+            n_samples,
+            memory=memory,
+            memory_level=memory_level,
+            random_state=random_state
+        ) for img, confound, n_samples in zip(imgs, confounds,
+                                                      subject_n_samples))
+
+    current_position = 0
+    for i, next_position in enumerate(np.cumsum(subject_n_samples)):
```

Comment:
what about:

current_pos = 0
```python
for data in data_list:
    len_data = len(data)
    data[current_pos:current_pos+len_data] = data
    current_pos += len_data
```

At the end of the day, isn't this thing just a concatenation?